torch/fx/passes/runtime_assert.py

```diff
@@ -251,11 +257,21 @@ def match_symbol(symint, cb):
                     if (
                         isinstance(symint, torch.SymInt)
                         and isinstance(symint.node, SymNode)
-                        and isinstance(s := symint.node.expr, sympy.Symbol)
-                        and s not in expr_to_proxy
                     ):
-                        expr_to_proxy[s] = fx.Proxy(cb())
-                        log.debug("expr_to_proxy[%s] = %s", s, expr_to_proxy[s])
+                        s = symint.node.expr
+                        if isinstance(s, sympy.Symbol) and s not in expr_to_proxy:  # basic symbol
+                            expr_to_proxy[s] = fx.Proxy(cb())
+                            log.debug("expr_to_proxy[%s] = %s", s, expr_to_proxy[s])
+                        elif (
+                            isinstance(s, sympy.Expr)
+                            and shape_env.prefer_deferred_runtime_asserts_over_guards
+                            and s in replacement_expr_to_symbol
+                            and (_s := replacement_expr_to_symbol[s]) not in expr_to_proxy
+                        ):  # symbol replaced with expression
+                            # if we've set a replacement, e.g. s2 = s0*s1, s0*s1 shows up as the shape,
+                            # so we store expr_to_proxy[s2] = cb(), and we're able to reify the assert.
+                            expr_to_proxy[_s] = fx.Proxy(cb())
+                            log.debug("expr_to_proxy[%s] = %s", _s, expr_to_proxy[_s])
```

Comment:
This doesn't feel tight. If you want, I'll think more about what a tight way of doing this, but this is probably what I'd think about on how to express it in a more general way.